app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Validator.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2017 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.sharedmodel.formtemplate
+
+import cats.data.Validated
+import cats.data.Validated.{ Invalid, Valid }
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+import uk.gov.hmrc.gform.core.parsers.ExprParsers
+import uk.gov.hmrc.play.http.HeaderCarrier
+import play.api.libs.functional.syntax._
+import play.api.libs.functional.syntax._
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+sealed trait Validator {
+  def errorMessage: String
+  def validate(data: Map[FieldId, Seq[String]])(f: ((String, String)) => Future[Boolean])(implicit ec: ExecutionContext, hc: HeaderCarrier): Future[Validated[Map[FieldId, Set[String]], Unit]]
```

Comment:
Here we're breaking _Single Responsibility Principle_. `Validator` suppose to be a part of a template, just plain data structure, kind of instruction for later interpretation. Interpretation and therefore validating  should happen somwhere else. 
How about moving this method to somewhere else (maybe to `ValidationService`)?

Reply:
I thought about that. however it is called inside the frontend validation service and the desired function is passed in. its cause each validator may have different implementation depending on what needs to be done on it. and it makes more sense to me to add it to this object.